From 196d95d2d225a7adcec5d7ee4462c5db6ad8528f Mon Sep 17 00:00:00 2001
From: Bibo Mao <maobibo@loongson.cn>
Date: Fri, 6 Jun 2025 14:30:26 +0800
Subject: [PATCH 203/285] hw/intc/loongarch_pch_msi: Inject MSI interrupt to
 kernel

If kvm_irqchip_in_kernel() return true, MSI interrupt can be injected
with API kvm_irqchip_send_msi() to KVM.

Reviewed-by: Song Gao <gaosong@loongson.cn>
Signed-off-by: Bibo Mao <maobibo@loongson.cn>
Message-ID: <20250606063033.2557365-7-maobibo@loongson.cn>
Signed-off-by: Song Gao <gaosong@loongson.cn>
Signed-off-by: Xianglai Li <lixianglai@loongson.cn>
Signed-off-by: Binbin Zhou <zhoubinbin@loongson.cn>
---
 hw/intc/loongarch_pch_msi.c | 34 ++++++++++++++++------------------
 1 file changed, 16 insertions(+), 18 deletions(-)

diff --git a/hw/intc/loongarch_pch_msi.c b/hw/intc/loongarch_pch_msi.c
index 81593333ce85..9fcdf44e5bba 100644
--- a/hw/intc/loongarch_pch_msi.c
+++ b/hw/intc/loongarch_pch_msi.c
@@ -2,7 +2,7 @@
 /*
  * QEMU Loongson 7A1000 msi interrupt controller.
  *
- * Copyright (C) 2024 Loongson Technology Corporation Limited
+ * Copyright (C) 2021 Loongson Technology Corporation Limited
  */
 
 #include "qemu/osdep.h"
@@ -13,9 +13,8 @@
 #include "hw/pci/msi.h"
 #include "hw/misc/unimp.h"
 #include "migration/vmstate.h"
-#include "trace.h"
 #include "sysemu/kvm.h"
-#include "hw/loongarch/virt.h"
+#include "trace.h"
 
 static uint64_t loongarch_msi_mem_read(void *opaque, hwaddr addr, unsigned size)
 {
@@ -28,24 +27,23 @@ static void loongarch_msi_mem_write(void *opaque, hwaddr addr,
     LoongArchPCHMSI *s = (LoongArchPCHMSI *)opaque;
     int irq_num;
 
-    MSIMessage msg = {
-        .address = addr,
-        .data = val,
-    };
+    if (kvm_irqchip_in_kernel()) {
+        MSIMessage msg;
 
-    if (kvm_enabled() && kvm_irqchip_in_kernel()) {
+        msg.address = addr;
+        msg.data = val;
         kvm_irqchip_send_msi(kvm_state, msg);
-    } else {
-        /*
-         * vector number is irq number from upper extioi intc
-         * need subtract irq base to get msi vector offset
-         */
-        irq_num = (val & 0xff) - s->irq_base;
-        trace_loongarch_msi_set_irq(irq_num);
-        assert(irq_num < s->irq_num);
-
-        qemu_set_irq(s->pch_msi_irq[irq_num], 1);
+        return;
     }
+
+    /*
+     * vector number is irq number from upper extioi intc
+     * need subtract irq base to get msi vector offset
+     */
+    irq_num = (val & 0xff) - s->irq_base;
+    trace_loongarch_msi_set_irq(irq_num);
+    assert(irq_num < s->irq_num);
+    qemu_set_irq(s->pch_msi_irq[irq_num], 1);
 }
 
 static const MemoryRegionOps loongarch_pch_msi_ops = {
-- 
2.51.0

